src/cos.py

```diff
@@ -0,0 +1,240 @@
+#  Copyright 2023 Canonical Ltd.
+#  See LICENSE file for licensing details.
+
+"""The COS integration observer."""
+import json
+import logging
+
+import ops
+from charms.loki_k8s.v0.loki_push_api import (
+    LokiPushApiConsumer,
+    LokiPushApiEndpointDeparted,
+    LokiPushApiEndpointJoined,
+)
+from ops import EventBase, Relation
+from pydantic import AnyHttpUrl, BaseModel
+
+import promtail
+from charm_state import State
+from event_timer import EventTimer
+
+METRICS_LOGGING_INTEGRATION_NAME = "metrics-logging"
+
+PROMTAIL_HEALTH_CHECK_INTERVAL_MINUTES = 5
+
+logger = logging.getLogger(__name__)
+
+
+class PromtailBinary(BaseModel):
```

Comment:
IMO, these whould be in the loki lib

Reply:
@merkata (cc @arturo-seijas) Maybe you can address this in today's COS meeting: Would it be possible for the lib to provide dataclasses/pydantic models with appropriate validation? So in principle this would mean that the 3 dataclasses here and the method `_retrieve_loki_integration_data` would be in the `loki_push_api` lib.

Reply:
> @merkata (cc @arturo-seijas) Maybe you can address this in today's COS meeting: Would it be possible for the lib to provide dataclasses/pydantic models with appropriate validation? So in principle this would mean that the 3 dataclasses here and the method `_retrieve_loki_integration_data` would be in the `loki_push_api` lib.

defo, thanks for raising @cbartz  (and @arturo-seijas )!

Reply:
> > @merkata (cc @arturo-seijas) Maybe you can address this in today's COS meeting: Would it be possible for the lib to provide dataclasses/pydantic models with appropriate validation? So in principle this would mean that the 3 dataclasses here and the method `_retrieve_loki_integration_data` would be in the `loki_push_api` lib.
> 
> defo, thanks for raising @cbartz (and @arturo-seijas )!

@sed-i  @0x12b fyi

Reply:
I have removed the use of Promtail. Will resolve the conversation.